Add `haskell.lib.incremental` utility

Gabriella439 opened the pull request:
This adds a new `incremental` utility for Haskell CI that supports incremental builds based on the approach outlined in this blog post:

https://harry.garrood.me/blog/easy-incremental-haskell-ci-builds-with-ghc-9.4/

The basic idea is that instead of Nix doing a full build for a package, we split every build into two builds:

- A full build at an older point in time

  e.g. a daily or weekly time boundary

- An incremental build relative to the last full build

  This incremental build reuses the build products left over from the most recent full build.

In order to do this, though, we need a way to "snap" a package's `git` source input to an earlier point in time (e.g. a daily boundary or weekly boundary).  This would allow multiple incremental builds to share the same full rebuild if they snap to the same time boundary.

The approach I went with to make that possible was to extend Nix's `builtins.fetchGit` to support a new `date` argument and you can find the corresponding PR for that here:

https://github.com/NixOS/nix/pull/7362

That is why the `incremental` utility added here requires a sufficiently new version of Nix (one that would incorporate that change, presuming it is merged).

This also requires GHC 9.4 or newer in order to pick up a fix to GHC's change detection logic, as described in  more detail in the above blog post.

However, if you satisfy those requirements then this works exactly the way you'd expect: all of the incremental builds only have to build the diff since the last time boundary. Moreover, if CI caches the full build then developers can also run `nix build` locally and only have to build the diff, too.

iteratee committed:
haskell.compiler.ghc92*: backport Cabal bugfix for Paths_ modules

There is a code generation bug in Cabal-3.6.3.0. For packages configured with
--enable-relocatable, Cabal would generate code that doesn't compile.

There isn't an upstream issue, but the issue is described in the commit that
fixed it:
https://github.com/haskell/cabal/commit/6c796218c92f93c95e94d5ec2d077f6956f68e98

It was fixed in Cabal-3.8.*
Backport the fix to the Cabal library that ships with ghc-9.2.4

Co-authored-by: sternenseemann <sternenseemann@systemli.org>

Closes #200063.

ncfavier committed:
lambdabot: remove upstreamed patches

cdepillabout committed:
haskellPackages: stackage LTS 19.32 -> LTS 20.1

This commit has been generated by maintainers/scripts/haskell/update-stackage.sh

cdepillabout committed:
all-cabal-hashes: 2022-11-11T17:48:48Z -> 2022-11-20T07:33:12Z

This commit has been generated by maintainers/scripts/haskell/update-hackage.sh

cdepillabout committed:
haskellPackages: regenerate package set based on current config

This commit has been generated by maintainers/scripts/haskell/regenerate-hackage-packages.sh

zowoq committed:
Revert "haskellPackages.hercules-ci-agent: add patch"

This reverts commit 34f6edc870a420f1edc0b32cf85b8b5f3ee2949c.

patch is included in 0.9.8

ncfavier committed:
haskellPackages: fixes

ncfavier committed:
haskellPackages: bump default GHC to 9.2.x

thomas-huegel committed:
morpheus-graphql not broken

sternenseemann committed:
hpack: work around test suite choking on CRLF

ncfavier committed:
haskellPackages.oeis: jailbreak

cdepillabout committed:
haskellPackages.vinyl: remove unnecessary override

cdepillabout committed:
haskellPackages.aeson-diff: remove unnecessary jailbreak

cdepillabout committed:
haskellPackages.heist: remove unnecessary overrides

cdepillabout committed:
haskellPackages.flat: remove unnecessary overrides

cdepillabout committed:
haskell.packages.ghc94.doctest: don't force specific version

cdepillabout committed:
haskellPackages.linear-base: remove unnecessary overrides

cdepillabout committed:
haskell.packages.ghc94.invariant: remove jailbreak

cdepillabout committed:
haskell.packages.ghc94.implicit-hie-cradle: remove jailbreak

cdepillabout committed:
haskell.packages.ghc94.tasty-hedgehog: override version 1.3.1.0 with 1.4.0.0

cdepillabout committed:
haskell.packages.ghc94.retry: remove dontCheck

cdepillabout committed:
haskell.packages.ghc94.rerebase: remove jailbreak

cdepillabout committed:
haskell.packages.ghc94.hashtables: remove jailbreak

cdepillabout committed:
haskellPackages.lens: ship lens 5.1.1 which is from stackage

cdepillabout committed:
haskellPackages.swagger2: dont jailbreak

cdepillabout committed:
haskellPackages.servant-*: remove doJailbreak

cdepillabout committed:
haskellPackages.servant-swagger: dont jailbreak

cdepillabout committed:
haskell.packages.ghc94.swagger2: dont jailbreak

cdepillabout committed:
haskell.packages.ghc94.lens-aeson: remove unnecessary override

sternenseemann committed:
haskell.packages.ghc92.ghc-lib-parser-ex: use stackage version

This unbreaks hlint which constrains it to < 9.2.1.

sternenseemann committed:
haskell.packages.ghc8{8,10}.base-compat-batteries: add missing dep

sternenseemann committed:
haskellPackages.OrderedBits: restrict to x86 platforms

Suggested in https://github.com/NixOS/nixpkgs/pull/202240.

sternenseemann committed:
haskellPackages.mbox: fix build for time >= 1.10

Patch proposed to upstream via email, using a trivial replacement
as suggested in the time documentation.

sternenseemann committed:
haskellPackages: adopt maintainership of core haskell infra pkgs

peti is no longer active and I am working on these packages anyways /
am upstream maintainer anyways.

sternenseemann committed:
haskellPackages.{pandoc,hlegder}: adopt maintainership from peti

Commit:
Add `haskell.lib.incremental` utility

This adds a new `incremental` utility for Haskell CI that
supports incremental builds based on the approach outlined
in this blog post:

https://harry.garrood.me/blog/easy-incremental-haskell-ci-builds-with-ghc-9.4/

The basic idea is that instead of Nix doing a full build
for a package, we split every build into two builds:

- A full build at an older point in time

  e.g. a daily or weekly time boundary

- An incremental build relative to the last full build

  This incremental build reuses the build products left
  over from the most recent full build.

In order to do this, though, we need a way to "snap" a
package's `git` source input to an earlier point in time
(e.g. a daily boundary or weekly boundary).  This would
allow multiple incremental builds to share the same
full rebuild if they snap to the same time boundary.

The approach I went with to make that possible was to
extend Nix's `builtins.fetchGit` to support a new `date`
argument and you can find the corresponding PR for that
here:

https://github.com/NixOS/nix/pull/7362

That is why the `incremental` utility added here requires
a sufficiently new version of Nix (one that would incorporate
that change, presuming it is merged).

This also requires GHC 9.4 or newer in order to pick up a fix
to GHC's change detection logic, as described in  more detail
in the above blog post.

However, if you satisfy those requirements then this works
exactly the way you'd expect: all of the incremental builds
only have to build the diff since the last time boundary.
Moreover, if CI caches the full build then developers can
also run `nix build` locally and only have to build the diff, too.

sternenseemann committed:
haskellPackages: maralorn adopts packages from peti

arcz committed:
haskellPackages: remove unnecessary overrides from ghc-9.2.x

arcz committed:
haskellPackages: configuration cleanup

sternenseemann committed:
haskellPackages: update base compiler and core pkgs in hackage2nix cfg

This is important for solving cabal flags and constraints when generating
the main haskell package set and its package definitions. After updating
the definition, more packages should be generated correctly for GHC 9.2,
e.g. ghc-exactprint.

sternenseemann committed:
haskellPackages.wide-word: lift constraint on hedgehog

sternenseemann committed:
haskellPackages.type-errors: apply upstream patch for 9.2

Commit:
haskell: Fix typo

Commit:
haskellPackages.monad-bayes: Mark unbroken

Commit:
haskellPackages: Sort and extend my maintainership

ncfavier committed:
haskell.packages.ghc9{2,4}.haskell-language-server: remove plugin overrides

Now that the base compiler has been updated to GHC 9.2, the cabal file
removes those unsupported plugins.

ncfavier committed:
haskellPackages.haskell-language-server: add missing dep

This is needed because hackage2nix doesn't know we want to disable the
`auto` flag, so it adds the wrong dependencies.

sternenseemann committed:
haskellPackages.singleton-th: allow th-desugar-1.14

sternenseemann committed:
haskellPackages.multipart: allow bytestring-0.11

sternenseemann committed:
haskellPackages.tasty-discover: disable broken test suite

maralorn committed:
haskell-language-server: Update wrapper to ghc 92 and don‘t provide version less executable

sternenseemann committed:
haskellPackages.irc-{client,conduit}: lift upper bound on time and bytestring

sternenseemann committed:
haskellPackages.hashing: allow bytestring-0.11.*

sternenseemann committed:
haskellPackages.hnix: 0.14.0.8 -> 0.16

0.14 doesn't seem to be buildable with LTS 20 / GHC 9.2, so we'll have to
try and upgrade. I've also downgraded hnix-store* globally mostly because
it was kind of silly to have them on latest before – those packages are
not really used independently of hnix.

sternenseemann committed:
haskellPackages.dhall-nix: 1.1.24 -> 1.1.25

This matches the dhall version shipped by stackage. Note that it
still doesn't build due to hnix being broken at the moment.

cdepillabout committed:
haskellPackages.hspec-contrib: dontCheck

cdepillabout committed:
stack: remove overrides

cdepillabout committed:
haskell.packages.ghc90.hashable: add missing dep of base-orphans

cdepillabout committed:
haskell.packages.ghc90.hashable: change addExtraLibrary to addBuildDepend

cdepillabout committed:
spago: build with ghc90 instead of ghc92

cdepillabout committed:
haskell.packages.ghc810.hashable: add build depend on base-orphans

cdepillabout committed:
haskell.packages.ghc88.hashable: add build depend on base-orphans

cdepillabout committed:
spago: mark broken in ghc92

sternenseemann committed:
haskellPackages.Unique: allow hashable 1.4.*

sternenseemann committed:
haskellPackages.rope-utf16-splay: downgrade to match text version

sternenseemann committed:
haskell-ci: allow Stackage LTS-20 dependency version

sternenseemann committed:
all-cabal-hashes: 2022-11-20T07:33:12Z -> 2022-11-26T14:57:41Z

This commit has been generated by maintainers/scripts/haskell/update-hackage.sh
and maintainers/scripts/haskell/regenerate-hackage-packages.nix

sternenseemann committed:
haskellPackages.text-format: drop upstreamed patch

sternenseemann committed:
haskell.packages.ghc8107.purescript-*st: mark as broken

These packages are not really used by anything anymore (except dovetail
which has not been working (ever?)). As they now choke on the
base-compat version we ship, we'll mark them as broken.

sternenseemann committed:
haskellPackages.purescript: drop unnecessary overrides

We no longer need to fiddle around with the broken flag, as the package
is in the default set now.

sternenseemann committed:
haskellPackages.purenix: adjust override for pure purescript dep

Before, purenix relied on purescript-*st which is no longer the case. It
now depends on the main purescript package and needs to inherit its
overrides.

sternenseemann committed:
haskellPackages: move purescript override block into inherit

Since the “purescript block” was at the end of the file before (after a
set update), a lot of unrelated overrides where added to it
accidentally, leading to a confusing collection of overrides in there.
With the inherit method of establishing a shared scope for a handful of
overrides this should no longer be a problem.

sternenseemann committed:
hledger-check-fancyassertions: update source hash for 1.27.1

arcz committed:
haskellPackages: configuration common cleanup

arcz committed:
haskellPackages: unbreak selected packages

cdepillabout committed:
haskellPackages.vinyl-updates: remove unused override

cdepillabout committed:
haskellPackages.htoml: mark broken

arcz committed:
haskellPackages: unbreak selected packages

Gabriella439 commented:
Needless to say, this should not be merged until https://github.com/NixOS/nix/pull/7362 is merged, but I'm putting up this PR for review anyway since it informs the upstream PR to Nix

Commit:
Lower required version

… so that it works against the upstream PR

I'll change the required version to an official release
if the PR is merged.

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -79,6 +79,7 @@ in
 , enableSeparateBinOutput ? false
 , enableSeparateDataOutput ? false
 , enableSeparateDocOutput ? doHaddock
+, enableSeparateDistOutput ? false
```
I guess this is slightly different  than options like `enableSeparateBinOutput`, `enableSeparateDataOutput`, etc.

Even if  `enableSeparateBinOutput` and `enableSeparateDataOutput` are set to `false`, the binary and data files still get into the single, main output.  But if `enableSeparateDistOutput` is `false`, then the `dist/build/` directory just completely doesn't get output at all.

I think this is fine as-is (and is likely what most users would expect to happen), but maybe you could add a comment here just clarifying this?

sternenseemann replied:
I think it should be renamed. The other ones cause something that would be installed anyways to go to a separate output. This flag moves something that would not have been "installed" otherwise to be moved into an output.

Maybe something starting with `do`? Although there ought to be something less awkward and ambiguous than `doInstallDist`…

Gabriella439 replied:
If we were to go the `doInstallDist` route, then for consistency we would want to add both `doInstallDist` *and* `enableSeparateDistOutput`.  I can make that change, but I just wanted to check if that would be alright with you two

cdepillabout commented on pkgs/development/haskell-modules/lib/compose.nix:
```diff
@@ -463,4 +463,59 @@ rec {
+  # incremental build reuses the `dist` directory from the last full rebuild.
+  #
+  # This only works for packages that use `git` for their source.
+  incremental = { interval }: pkg:
```
Could you give one or two examples of what sorts of values the `interval` argument should take in the above comment?

cdepillabout commented on pkgs/development/haskell-modules/lib/default.nix:
```diff
@@ -354,4 +354,6 @@ rec {
   # built. Will delay failures, if any, to compile time.
   allowInconsistentDependencies = compose.allowInconsistentDependencies;
+
+  incremental = pkgs: args: compose.incremental args pkgs;
```
Very small nitpick, but `pkgs` is currently only a single package, right?

```suggestion
  incremental = pkg: args: compose.incremental args pkg;
```

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -705,6 +718,10 @@ stdenv.mkDerivation ({
 // optionalAttrs (args ? postFixup)              { inherit postFixup; }
 // optionalAttrs (args ? dontStrip)              { inherit dontStrip; }
+// optionalAttrs enableSeparateDistOutput {
+     # Don't strip the dist output
+     preFixupHooks = [ "outputs=(\${outputs[@]/dist})" ];
```
I'm purely just curious, but how does this prevent the `dist` output from getting stripped?  (I'm not familiar with `preFixupHooks`.)

cdepillabout committed:
haskellPackages.snaplet-sqlite-simple: unmark as broken

cdepillabout committed:
haskellPackages: regenerate

sternenseemann commented on pkgs/development/haskell-modules/lib/compose.nix:
```diff
@@ -463,4 +463,59 @@ rec {
+  # This only works for packages that use `git` for their source.
+  incremental = { interval }: pkg:
+    let
+      requiredNixVersion = "2.12.0pre20221128_32c182b";
```
Please remove this function from the PR. If I understand it correctly, it depends on https://github.com/NixOS/nix/pull/7362 – “presuming it is merged” is presumed too much. In general I also want to keep anything out of `haskell.lib` that depends on unreleased Nix features.

Presumed too much also in the sense that I believe the `fetchGit` feature will be quite controversial.

Gabriella439 committed:
s/pkgs/pkg/g

… as caught by @cdepillabout

Co-authored-by: Dennis Gosnell <cdep.illabout@gmail.com>

Commit:
Skip the use of `tar`

We can store the `dist` directory decompressed, which
speeds up the dist export/import

This potentially requires more disk space *but* by storing
the files unpacked it may actually improve disk utilization
in some cases if `auto-optimise-store` is enabled by
permitting deduplication of `dist` files.

Commit:
Add an `installDist` phase

… which is disabled by default

The motivation for this is to bring the behavior of
`enableSeparateDistOutput` more in line with the other
options where it doesn't change *whether* or not something
is exported, but rather *where* it is exported.

Now `installDist` controls whether or not the `dist`
directory is exported.

Based on this discussion:

https://github.com/NixOS/nixpkgs/pull/203499/files#r1034150076

sternenseemann committed:
dhall-nix: patch for compatibility with hnix 0.16

sternenseemann committed:
dhall-nixpkgs: allow building with hnix 0.16

Commit:
Document `interval` argument

… as suggested by @cdepillabout

Commit:
s/for use for/for use with/

… based on feedback from @MaxGabriel

Commit:
Move `installDistPhase` to `postPhases`

There are two reasons for doing this:

- We can get rid of the hack to remove the dist output from the outputs

- We can ensure that any changes that happen in the install phase are
  correctly reflected in the `dist` export

Gabriella439 commented:
I rebased this on top of #203963 because the `.dylib` workaround interferes with the incremental build.  Specifically, the `package.conf` surgery entailed by the `.dylib` workaround causes downstream modules to be rebuilt due to `Cabal`'s change detection

Commit:
Disable dylib workaround for incremental build

Gabriella439 commented:
Sorry for the unnecessary pings.  This was a bad rebase on my part.  I'll fix this
